### ISSUE TITLE
Deprecated ignored if `since` > rust version and has patch

### ISSUE THREAD

**Issue by ascjones** [rust-lang/rust#57952]
When `since` > version of `rustc` *and* patch is in the version then deprecated is ignored:

My expectation is that: `since` is referring to the version of my crate, **not** the rustc version.

```rust
struct Bar {}
impl Bar {
    #[deprecated(since = "1.32.0")]
    fn patch_works(&self) -> () { () }
    
    #[deprecated(since = "1.33.0")]
    fn patch_ignored1(&self) -> () { () }
    
    #[deprecated(since = "2.1.0")]
    fn patch_ignored2(&self) -> () { () }
    
    #[deprecated(since = "2.0")]
    fn minor_works(&self) -> () { () }
}

fn main() {
    let bar = Bar { };
    bar.patch_works();
    bar.patch_ignored1();
    bar.patch_ignored2();
    bar.minor_works();
}
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=99412d4ad3fa14e5c5b0a5070ee16b65))

The [docs](https://doc.rust-lang.org/reference/attributes.html#deprecation) say: 

> since expects a version number, as in #[deprecated(since = "1.4.1")]
> rustc doesn't know anything about versions, but external tools like clippy may check the validity of this field.

**Comment by varkor** [comment]
Version numbers are according to [semantic versioning](https://semver.org/), in which there must be three components. `2.0` is not a valid version number and as such as ignored.

**Comment by varkor** [comment]
If you add `#![warn(deprecated_in_future)]` to the example (on Beta or Nightly), then it will warn for future versions too.

**Comment by varkor** [comment]
> it doesn't really make sense to deprecate things in a crate based on the Rust version

Ah, that's true. Maybe it does make more sense to always warn in this case.

**Comment by ascjones** [comment]
> it doesn't really make sense to deprecate things in a crate based on the Rust version

Exactly. My understanding from the [docs for the deprecation attribute](https://doc.rust-lang.org/reference/attributes.html#deprecation) is that `since` should refer to the version of *my* crate - nothing to do with the version of `rustc`. 

Sorry for not being clear enough originally - the problem was I was troubleshooting why a `#[deprecated]` attribute I had applied was being ignored and I narrowed it down to the cases in the example - and that's when I noticed that it seemed to relate to the `rustc` version, which looks like a bug, and decided to report it.

I will update the description
